File: packages/server/src/player.ts
/**
 * Player store — persistent player identity (survives avatar lifecycle).
 *
 * A PlayerRecord is created at registration (REST API) and persists
 * for the lifetime of the server process (in-memory for now).
 * The token is the player's session cookie — used to authenticate
 * WebSocket connections and reconnect to existing avatars.
 *
 * This is the same whether the player is a human or an AI agent;
 * registration is just an API call.
 */

import { randomUUID } from "node:crypto";

export interface PlayerRecord {
  /** Persistent session token (UUID) — stored in client cookie */
  token: string;
  /** Display name */
  name: string;
  /** Character definition ID (for appearance) */
  characterId: string;
  /** When this player registered */
  createdAt: number;
}

export interface PlayerStore {
  /** Register a new player, returns the created record */
  register(name: string, characterId: string): PlayerRecord;
  /** Look up a player by token, or null if not found */
  getByToken(token: string): PlayerRecord | null;
}

export class MemoryPlayerStore implements PlayerStore {
  private players = new Map<string, PlayerRecord>();

  register(name: string, characterId: string): PlayerRecord {
    const token = randomUUID();
    const record: PlayerRecord = {
      token,
      name,
      characterId,
      createdAt: Date.now(),
    };
    this.players.set(token, record);
    console.log(`[PlayerStore] Registered "${name}" (char: ${characterId})`);
    return record;
  }

  getByToken(token: string): PlayerRecord | null {
    return this.players.get(token) ?? null;
  }
}

File: game/server/internal/game/player.go
package game

import (
	"sync"

	"github.com/google/uuid"
)

// PlayerRecord represents a registered player.
// Created at registration (REST API), persists for the server's lifetime.
type PlayerRecord struct {
	Token       string
	Name        string
	CharacterID string
	CreatedAt   int64 // unix millis
}

// PlayerStore manages player registrations. Thread-safe via RWMutex.
// HTTP handlers call Register/GetByToken, then the token flows into the World
// via WebSocket join messages. The World's Run goroutine also calls GetByToken
// when handling join messages, so the mutex is needed.
type PlayerStore struct {
	mu      sync.RWMutex
	players map[string]*PlayerRecord // token → record
}

// NewPlayerStore creates an empty player store.
func NewPlayerStore() *PlayerStore {
	return &PlayerStore{
		players: make(map[string]*PlayerRecord),
	}
}

// Register creates a new player with a unique token.
func (s *PlayerStore) Register(name, characterID string) *PlayerRecord {
	token := uuid.New().String()
	rec := &PlayerRecord{
		Token:       token,
		Name:        name,
		CharacterID: characterID,
		CreatedAt:   nowMillis(),
	}
	s.mu.Lock()
	s.players[token] = rec
	s.mu.Unlock()
	return rec
}

// GetByToken looks up a player by token, or nil if not found.
func (s *PlayerStore) GetByToken(token string) *PlayerRecord {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.players[token]
}
